src/api/src/main/java/org/locationtech/geogig/model/Node.java

```diff
@@ -321,12 +330,37 @@ public void expand(Envelope env) {
         }
     }
 
-    @Nullable
+    /**
+     * Returns the extra data stored with this node, which shall be considered immutable, regardless
+     * of the concrete Map implementation returned.
+     * <p>
+     * Changes to the returned map, if it's not immutable, will not affect the internal state of the
+     * node.
+     * <p>
+     * The returned map may contain {@code null} values, but not {@code null} key.
+     * 
+     * @return a non-null, possibly empty <b> copy </b> of this node's extra data map
+     */
     public Map<String, Object> getExtraData() {
-        return extraData;
+        return convertNullObjectValueToNull(extraData);
```

Comment:
Another option would be for the internal state to be a HashMap instead of an ImmutableMap, but that would imply having to return a "safe copy", which is much more expensive than a transform